### ISSUE TITLE
Request more information from zcAjaxPayment

### ISSUE THREAD

**Issue by mc12345678** [zencart/zencart#4201]
**Is your feature request related to a problem? Please describe.**

Working with Stock By Attributes, if a purchase is made using zcAjaxPayment, then there isn't enough information necessarily to pass along to determine the extent of the `$order`. I'm thinking that just before the `for` loop associated with stock that in some way the `$order` be shared for "reading"/"retrieval" from outside of the `prepareConfirmation` method.  Based on the program flow it looks like at least for SBA, if the `$order` can be stored as a variable of the class, that there is a variable that is generated in support of the ajax that can be referenced by the follow on code in the global space that would allow accessing the data as built/modified along the way including from order total modules, etc..,, 

Cleaner would be to support transfer of the `$order` value just before use, then to transfer the current value along with the `$i` within the loop so that each value can independently be reviewed and/or provided an independent response.

I could duplicate the class file to be overridden/replace the core store version, one way I could modify the code is to adjust the portion that sends the `$products_id` to be an array containing at least two keys: `products_id` and `attributes`. The software only needs to understand/manage combination of products_id and `attributes` to be able to identify a product.

````php
    $flagAnyOutOfStock = false;
    $stock_check = array();
    if (STOCK_CHECK=='true') {
      for($i = 0, $n = sizeof ($order->products); $i<$n; $i++) {
        if ($stock_check[$i] = zen_check_stock ($order->products[$i]['id'], $order->products[$i]['qty'])) {
          $flagAnyOutOfStock = true;
        }
      }
      // Out of Stock
      if ((STOCK_ALLOW_CHECKOUT!='true')&&($flagAnyOutOfStock==true)) {
        zen_redirect (zen_href_link (FILENAME_SHOPPING_CART));
      }
    }
````

Specifically interested in this portion of the code:
````php
        if ($stock_check[$i] = zen_check_stock ($order->products[$i]['id'], $order->products[$i]['qty'])) {
````
where one way it could be improved for current use is:
````diff
-        if ($stock_check[$i] = zen_check_stock ($order->products[$i]['id'], $order->products[$i]['qty'])) {
+        if ($stock_check[$i] = zen_check_stock (array('products_id' => $order->products[$i]['id'], 'attributes' => $order->products[$i]['attributes']), $order->products[$i]['qty'])) {
````
That is one of any many potential solutions.

````diff
      for($i = 0, $n = sizeof ($order->products); $i<$n; $i++) {
+        $GLOBALS['zco_notifier']->notify('NOTIFY_AJAX_PAYMENT_PREPARE_CONFIRMATION', null, $order, $i);
        if ($stock_check[$i] = zen_check_stock ($order->products[$i]['id'], $order->products[$i]['qty'])) {
          $flagAnyOutOfStock = true;
        }
      }
````

````diff
    if (STOCK_CHECK=='true') {
+      $this->order = $order
      for($i = 0, $n = sizeof ($order->products); $i<$n; $i++) {
+        $GLOBALS['zco_notifier']->notify('NOTIFY_AJAX_PAYMENT_PREPARE_CONFIRMATION', $i);
        if ($stock_check[$i] = zen_check_stock ($order->products[$i]['id'], $order->products[$i]['qty'])) {
          $flagAnyOutOfStock = true;
        }
      }
````

Considered the possibility of just using a notifier in the area; however, `$order` is a local variable not a class variable and it needs to remain a local variable until everything has been sorted because the other software expects to find it as a local variable. 

There doesn't appear to be any way to retrieve internal variables of a function/method from downstream if they are not directly shared with the other code. 

Effectively trying to find a way to obtain back from the store, information pertinent to the specific product as could be retrieved through the SBA observer method: `setCheckStockParams`found at: https://github.com/mc12345678/Stock_By_Attributes_Combined/blob/cf1bd9a6e2ad3aacc2831494b34c531a6eb2df07/includes/classes/observers/class.products_with_attributes_stock.php#L1671

**Comment by drbyte** [comment]
> Cleaner would be to support transfer of the `$order` value just before use, then to transfer the current value along with the `$i` within the loop so that each value can independently be reviewed and/or provided an independent response.

FYI: Careful. Some refactorings are occurring to use `foreach` instead of the (usually) unnecessarily long and verbose `for($i = 0, $n = sizeof ($order->products); $i<$n; $i++) {`. So forcing things into being dependent on `$i` makes things unpleasantly brittle. In fact, old notifier hooks using $i are what's holding back some of these code improvements.

> There doesn't appear to be any way to retrieve internal variables of a function/method from downstream if they are not directly shared with the other code.

Correct. That's always intentional. Any hacks to overcome that should not be relied upon.

**Comment by drbyte** [comment]
> There doesn't appear to be any way to retrieve internal variables of a function/method from downstream if they are not directly shared with the other code.

Correct. That's always intentional. Any hacks to overcome that should not be relied upon.

**Comment by drbyte** [comment]
> Working with Stock By Attributes, if a purchase is made using zcAjaxPayment, then there isn't enough information necessarily to pass along to determine the extent of the $order

While the order class is already doing a zillion things, I wonder if it would be better to just be able to retrieve this stock information via the order class itself, where it populates product information?
I firmly believe that stock handling has nothing to do with an order, but when querying an order's details it makes sense to query stock status from one central point instead of the current state of having it initiated in a dozen places that all require customization when someone has a preferred alternate way of dealing with stock.

**Comment by drbyte** [comment]
Given that this section of code is mostly a clone of `/includes/modules/pages/checkout_confirmation/header_php.php`, how are you handling SBA needs when fired in that context instead of via ajax?

**Comment by mc12345678** [comment]
> Given that this section of code is mostly a clone of `/includes/modules/pages/checkout_confirmation/header_php.php`, how are you handling SBA needs when fired in that context instead of via ajax?

Well, in my "hack" of trying to address the last opportunity to identify that this purchase can be performed without maybe exceeding available stock, I have an observer awaiting the load of this page, that identifies when the quantity of stock is sought (via `zen_check_stock`) to effectively reach out to the global space to retrieve the attributes for the specific order item being evaluated based on also collecting `$i`. I am looking to go to a more generalized approach to handle the possibility of two or more `$order->products` elements having the same `products_id` and even the same other information.  That was part of the desired restructure but I didn't want to take it on without some semblance of what is/was to come.

It is at this point in the processing that appears to be the last opportunity to touch the product of the order so no matter what has been done "upfront" or before this point, this is when/where the "order" is complete.  If something downstream modifies the order, then that process needs to perform the necessary evaluation, but it shouldn't happen after this point.

Otherwise, things such as `order_totals`, `shipping`, `payment` methods etc... have the opportunity to modify the ordered items and there does not appear to be anything within those processes that show/identify a direct reference to the `$order` itself such that the current status is "known" at or just before the point in question.

Again, I currently use the provided `$i` in global space to support data retrieval from the also global space `$order`. 
I am trying to move forwards with a more "general" review of the `$order` such that don't have to rely on knowing `$i`, instead will know the `$products_id` and of the `$products_id` in the order, whether that specific `$products_id` or element had been previously evaluated. It seems that I likely will have to add a flag/variable/parameter to the `$order->products` class data to indicate status of evaluation rather than internally tracking.  This is because there are other outside processes that may modify the `$order->products` data in the loop throwing off my attempted evaluation. Fortunately, the data would only exist temporarily to support stock lookup.

As far as the order class doing a stock lookup? Sure, its just that I see from the path of variable creation to support placing the order that there are opportunities for the actual order to be modified after the class variable has been instantiated.

**Comment by drbyte** [comment]
You said "evaluation" several times. What does that entail? zen_check_stock adds a warning message if stock is insufficient, but doesn't prevent checkout. Are you just adding a warning message as well?

**Comment by mc12345678** [comment]
> You said "evaluation" several times. What does that entail? zen_check_stock adds a warning message if stock is insufficient, but doesn't prevent checkout. Are you just adding a warning message as well?

I have written a response to this more times than I wish to count or say. Effectively, the code that is used for the "standard" process of adding product to the cart determines if aspects of the added product are correct/acceptable through data analysis.  If not, then the product is not added and currently while the code permits modifying the default message, it is "setup" to basically provide the default message. Further it takes some "effort" to modify that message (which as historically noted includes html) if the default has been changed.

State that zen_check_stock adds a warning message, while this is true only in the situation where insufficient stock is identified, the presence of such message (regardless the content) does at least provide information in a default cart to indicate that checkout possibly should not occur.

The current proposed need (for evaluation) is to be able to know the details of the item being added to the cart to be able to basically determine if that item is a match against expected data to identify if the product in question is in stock.  While yes it would be helpful to have access to the message being returned, I do not currently see that as much as a necessity considering there may be other ways to present the message desired.  I realize that answer to that question may help identify the breadth of revision, but currently the access to the associated data seems to be limited.

Discussion or additional questions?

**Comment by mc12345678** [comment]
> FYI: Careful. Some refactorings are occurring to use `foreach` instead of the (usually) unnecessarily long and verbose `for($i = 0, $n = sizeof ($order->products); $i<$n; $i++) {`. So forcing things into being dependent on `$i` makes things unpleasantly brittle. In fact, old notifier hooks using `$i` are what's holding back some of these code improvements.

I've somewhat always wondered why this `for` loop had been implemented/retained as if it were in some way an expectation that the keys be sequentially identified with no gaps, etc... In light of the above comment, I have been changing the code in this plugin to use the `foreach` style while continuing to use `$i` as the key for the associated value. Is there some plan to remove reference to the key while looping through all product?! I mean, there are some benefits of using: `foreach ($array_list as &$value)` but, wouldn't the key still be desired to have/retain?

**Comment by drbyte** [comment]
> wouldn't the key still be desired to have/retain?

In some cases yes.
But with foreach syntax the key/counter would have to be manually incremented.
